File: runme/mcgen/local/datan/datsrc/mtxmar.f
      SUBROUTINE MTXMAR(A,B,ALAM,X1,X2,M,N,FRAC,OK)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      PARAMETER(NMAX=1000)
      DIMENSION A(M,N),B(M,1),X1(N),X2(N),D(NMAX),E(NMAX)
      LOGICAL OK
      COMMON /DASV01/ D,E
      PARAMETER(NB=1)
C STEP 1: Bidiagonalisation of A
      CALL MTXSV1(A,B,D,E,M,N,NB)
C STEP 2: Diagonalisation of bidiagonal matrix
      CALL MTXSV2(A,B,D,E,M,N,NB,OK)
C STEP 3: Order singular values and perform  permutations
      CALL MTXSV3(A,B,D,M,N,NB)
C STEP 4: Singular value analysis and appl. of Marquardt method
      CALL MTXSVM(A,B,D,ALAM,X1,X2,M,N,FRAC)
      END
C-----------------------------------------------------------------
      SUBROUTINE MTXSVM(A,B,D,ALAM,X1,X2,M,N,FRAC)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      PARAMETER(NMAX=1000)
      DIMENSION A(M,N),B(M),D(N),X1(N),X2(N),P1(NMAX),P2(NMAX)
      COMMON /DASV00/ P1,P2
      PARAMETER (EPSILN=1.D-15)
      FRACT=ABS(FRAC)
      IF(FRACT.LT.EPSILN) FRACT=EPSILN
      SINMAX=DBLE(0.)
      ALAM2=ALAM**2
      ALAMP=ALAM*DBLE(0.1)
      ALAMP2=ALAMP**2
      DO 10 I =  1, N
        SINMAX=MAX(SINMAX,D(I))
   10 CONTINUE
      SINMIN=SINMAX*FRACT
      KK=N
      DO 20 I=1,N
        IF(D(I).LE.SINMIN) THEN
          KK=I-1
          GO TO 30
        END IF
   20 CONTINUE
   30 CONTINUE
      DO 40 I =  1, M
        G=B(I)
        IF(I.LE.KK) THEN
          DEN1=1./(D(I)**2+ALAM2)
          DEN2=1./(D(I)**2+ALAMP2)
          P1(I)=G*D(I)*DEN1
          P2(I)=G*D(I)*DEN2
        ELSE
          IF(I.LE.N) THEN
            P1(I)=0.
            P2(I)=0.
          END IF
        END IF
   40 CONTINUE
      DO 60 I=1,N
        X1(I)=0.
        X2(I)=0.
        DO 50 K=1,N
          X1(I)=X1(I)+A(I,K)*P1(K)
          X2(I)=X2(I)+A(I,K)*P2(K)
   50   CONTINUE
   60 CONTINUE
      N=KK
      END

File: runme/mcgen/local/datan/datsrc/mtxmar.c
#include <stdlib.h>
#include "datsrc.h"

/* Common Block Declarations */

/*
struct {
    double d[1000], e[1000];
} dasv01_;

#define dasv01_1 dasv01_

struct {
    double p1[1000], p2[1000];
} dasv00_;

#define dasv00_1 dasv00_
*/

void LIBFUNC mtxmar_(double P_T *F_P(a), double P_T *b, double const P_T *alam,
             double P_T *x1, double P_T *x2, integer const P_T *m,
             integer P_T *n, double const P_T *frac, logical P_T *ok)
{
    /* Table of constant values */

    integer c__1 = 1;

    /* System generated locals */
    integer a_dim1, a_offset, b_dim1, b_offset;

    /* Local variables */
    double *d, *e;
    unsigned long amem;

#if defined(USEHUGE)
    HFP(double,a);
#endif

    *ok = FALSE_;
    if ((amem = sizeof(double) * 2 * *n) > MAXALLOC) {
        MEMERROR("mtxmar_ (too large)");
        goto f_0;
    }
    d = (double *)MEMALLOC(amem);
    if (!d) {
        MEMERROR("mtxmar_");
        goto f_0;
    }
    e = d + *n;

/* STEP 1: Bidiagonalisation of A */
    /* Parameter adjustments */
    b_dim1 = *m;
    b_offset = b_dim1 + 1;
    b -= b_offset;
    --x2;
    --x1;
    a_dim1 = *m;
    a_offset = a_dim1 + 1;
    a -= a_offset;

    /* Function Body */
    mtxsv1_((double P_T *)&a[a_offset], &b[b_offset], d, e, m, n, &c__1);
/* STEP 2: Diagonalisation of bidiagonal matrix */
    mtxsv2_((double P_T *)&a[a_offset], &b[b_offset], d, e, m, n, &c__1, ok);
/* STEP 3: Order singular values and perform  permutations */
    mtxsv3_((double P_T *)&a[a_offset], &b[b_offset], d, m, n, &c__1);
/* STEP 4: Singular value analysis and appl. of Marquardt method */
    mtxsvm_((double P_T *)&a[a_offset], &b[b_offset], d, alam, &x1[1], &x2[1],
           m, n, frac);
    MEMFREE((void *)d);
f_0:
    return;
} /* mtxmar_ */

/* ----------------------------------------------------------------- */
void LIBFUNC mtxsvm_(double const P_T *F_P(a), double const P_T *b,
             double const P_T *d, double const P_T *alam, double P_T *x1,
             double P_T *x2, integer const P_T *m, integer P_T *n,
             double const P_T *frac)
{
    /* System generated locals */
    integer a_dim1, a_offset, i__1, i__2;
    double d__1, d__2;

    /* Local variables */
    integer i, k, kk;
    double alam2, g, alamp, fract, alamp2, sinmin, sinmax, den1, den2;
    double *p1, *p2;
    unsigned long amem;

#if defined(USEHUGE)
    HFP(double,a);
#endif

    if ((amem = sizeof(double) * 2 * *m) > MAXALLOC) {
        MEMERROR("mtxmar_/mtxsvm_ (too large)");
        goto f_0;
    }
    p1 = (double *)MEMALLOC(amem);
    if (!p1) {
        MEMERROR("mtxmar_/mtxsvm_");
        goto f_0;
    }
    p2 = p1 + *m;

    /* Parameter adjustments */
    --b;
    --x2;
    --x1;
    --d;
    a_dim1 = *m;
    a_offset = a_dim1 + 1;
    a -= a_offset;

    /* Function Body */
    fract = abs(*frac);
    if (fract < 1e-15) {
        fract = 1e-15;
    }
    sinmax = 0.;
/* Computing 2nd power */
    d__1 = *alam;
    alam2 = d__1 * d__1;
    alamp = *alam * .1;
/* Computing 2nd power */
    d__1 = alamp;
    alamp2 = d__1 * d__1;
    i__1 = *n;
    for (i = 1; i <= i__1; ++i) {
/* Computing MAX */
        d__1 = sinmax, d__2 = d[i];
        sinmax = max(d__1,d__2);
    }
    sinmin = sinmax * fract;
    kk = *n;
    i__1 = *n;
    for (i = 1; i <= i__1; ++i) {
        if (d[i] <= sinmin) {
            kk = i - 1;
            goto L30;
        }
    }
L30:
    i__1 = *m;
    for (i = 1; i <= i__1; ++i) {
        g = b[i];
        if (i <= kk) {
/* Computing 2nd power */
            d__1 = d[i];
            den1 = 1.f / (d__1 * d__1 + alam2);
/* Computing 2nd power */
            d__1 = d[i];
            den2 = 1.f / (d__1 * d__1 + alamp2);
            p1[i - 1] = g * d[i] * den1;
            p2[i - 1] = g * d[i] * den2;
        } else {
            if (i <= *n) {
                p1[i - 1] = 0.f;
                p2[i - 1] = 0.f;
            }
        }
    }
    i__1 = *n;
    for (i = 1; i <= i__1; ++i) {
        x1[i] = 0.f;
        x2[i] = 0.f;
        i__2 = *n;
        for (k = 1; k <= i__2; ++k) {
            x1[i] += a[i + k * a_dim1] * p1[k - 1];
            x2[i] += a[i + k * a_dim1] * p2[k - 1];
        }
    }
    *n = kk;
    MEMFREE((void *)p1);
f_0:
    return;
} /* mtxsvm_ */

